cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,981 @@
+|string|`string_scalar`| This class object is immutable|
+
+### Construction
+`scalar`s can be created using either their respective constructors or using factory functions like 
+`make_numeric_scalar()`, `make_timestamp_scalar()` or `make_string_scalar()`. 
+
+### Casting
+All the factory methods return a `unique_ptr<scalar>` which needs to be statically downcasted to 
+its respective scalar class type before accessing its value. Their validity (nullness) can be 
+accessed without casting. Generally, the value needs to be accessed from a function that is aware 
+of the value type e.g. a functor that is dispatched from `type_dispatcher`. To cast to the 
+requisite scalar class type given the value type, use the mapping utility `scalar_type_t` provided 
+in `type_dispatcher.hpp` : 
+
+```c++
+//unique_ptr<scalar> s = make_numeric_scalar(...);
+
+using ScalarType = cudf::scalar_type_t<T>;
+// ScalarType is now numeric_scalar<T>
+auto s1 = static_cast<ScalarType *>(s.get());
+```
+
+### Passing to device
+Each scalar type has a corresponding non-owning device view class which allows access to the value 
+and its validity from the device. This can be obtained using the function 
+`get_scalar_device_view(ScalarType s)`. Note that a device view is not provided for a base scalar 
+object, only for the derived typed scalar class objects.
+
+# libcudf++ API and Implementation
+
+## Streams
+
+We do not yet expose CUDA streams in external libcudf APIs. 
+However, in order to ease the transition to future use of streams, all libcudf APIs that allocate device memory or execute a kernel should be implemented using asynchronous APIs on the default stream (e.g., stream 0). 
+
+The recommended pattern for doing this is to make the definition of the external API invoke an internal API in the `detail` namespace. The internal `detail` API will have all the same parameters, plus a `rmm::cuda_stream_view` parameter at the end defaulted to `rmm::cuda_stream_default`. 
+The implementation should be wholly contained in the `detail` API definition and use only asynchronous versions of CUDA APIs with the defaulted stream parameter. 
+
+In order to make the `detail` API callable from other libcudf functions, it should be exposed in a header placed in the `cudf/cpp/include/detail/` directory.
+
+For example:
+
+```c++
+// cpp/include/cudf/header.hpp
+void external_function(...);
+
+// cpp/include/cudf/detail/header.hpp
+namespace detail{
+void external_function(..., rmm::cuda_stream_view stream = rmm::cuda_stream_default)
+} // namespace detail
+
+// cudf/src/implementation.cpp
+namespace detail{
+    // defaulted stream parameter
+    void external_function(..., rmm::cuda_stream_view stream){
+        // implementation uses stream w/ async APIs
+        RMM_ALLOC(...,stream);
+        CUDA_TRY(cudaMemcpyAsync(...,stream.value()));
+        kernel<<<..., stream>>>(...);
+        thrust::algorithm(rmm::exec_policy(stream), ...);
+        stream.synchronize();
```

Comment:
This sync may lead people to believe that all detail functions should sync their stream before returning.